core/src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -263,11 +277,112 @@ public void connectToNode(DiscoveryNode node) throws ConnectTransportException {
         transport.connectToNode(node);
     }
 
-    public void connectToNodeLight(DiscoveryNode node) throws ConnectTransportException {
+    /**
+     * Lightly connect to the specified node, and handshake cluster
+     * name and version
+     *
+     * @param node             the node to connect to
+     * @param handshakeTimeout handshake timeout
+     * @return the connected node with version set
+     * @throws ConnectTransportException if the connection or the
+     *                                   handshake failed
+     */
+    public DiscoveryNode connectToNodeLight(final DiscoveryNode node, final long handshakeTimeout) throws ConnectTransportException {
+        return connectToNodeLight(node, handshakeTimeout, true);
+    }
+
+    /**
+     * Lightly connect to the specified node, returning updated node
+     * information. The handshake will fail if the cluster name on the
+     * target node mismatches the local cluster name and
+     * {@code checkClusterName} is {@code true}.
+     *
+     * @param node             the node to connect to
+     * @param handshakeTimeout handshake timeout
+     * @param checkClusterName whether or not to ignore cluster name
+     *                         mismatches
+     * @return the connected node
+     * @throws ConnectTransportException if the connection or the
+     *                                   handshake failed
+     */
+    public DiscoveryNode connectToNodeLight(final DiscoveryNode node, final long handshakeTimeout, final boolean checkClusterName) {
         if (node.equals(localNode)) {
-            return;
+            return localNode;
         }
         transport.connectToNodeLight(node);
+        try {
+            return handshake(node, handshakeTimeout, checkClusterName);
+        } catch (ConnectTransportException e) {
+            transport.disconnectFromNode(node);
+            throw e;
+        }
+    }
+
+    private DiscoveryNode handshake(
+            final DiscoveryNode node,
+            final long handshakeTimeout,
+            final boolean checkClusterName) throws ConnectTransportException {
+        final HandshakeResponse response;
+        try {
+            response = this.submitRequest(
+                node,
+                HANDSHAKE_ACTION_NAME,
+                new HandshakeRequest(),
+                TransportRequestOptions.builder().withTimeout(handshakeTimeout).build(),
```

Comment:
The liveness action is still needed because it's basically a health check *after* connection has taken place and is repeatedly periodically (default is every five seconds). However, it can be greatly simplified now so I pushed ccb32c21587b4f946a6a062d36cd6cb8693efc74. I also pushed 79b0da6b8b642d595d67368416268fe3a9499b17 to add the handshake action to the list of actions excluded from the transport tracer.